common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp

```diff
@@ -41,12 +63,136 @@ bool sta_wlan_hal_nl80211::connect(const std::string &ssid, const std::string &p
+                   << " on interface: " << get_iface_name();
+        return false;
+    }
+
+    LOG(ERROR) << "Network with id " << network_id << " has been added and enabled on interface "
+               << get_iface_name();
+
+    // Update active network parameters
+    m_active_ssid.assign(ssid);
+    m_active_bssid.assign(bssid);
+    m_active_pass.assign(pass);
+    m_active_channel    = channel;
+    m_active_network_id = network_id;
+
     return true;
 }
 
-bool sta_wlan_hal_nl80211::disconnect() { return true; }
+bool sta_wlan_hal_nl80211::disconnect()
+{
+    LOG(TRACE) << "Disconnect network id " << m_active_network_id
+               << " on interface: " << get_iface_name();
+
+    ConnectionStatus connection_status;
+    if (!read_status(connection_status)) {
+        LOG(ERROR) << "Failed reading status for " << get_iface_name() << "! can't disconnect";
+        return false;
+    }
 
-bool sta_wlan_hal_nl80211::roam(const sMacAddr &bssid, uint8_t channel) { return true; }
+    // Return gracefully if network state is not connected
+    if (!is_connected(connection_status.wpa_state)) {
+        LOG(DEBUG) << "Active network is not connected";
+        return true;
+    }
+
+    // Return gracefully if no network is connected
+    if (m_active_network_id < 0) {
+        LOG(DEBUG) << "Active network does not exist";
+        return true;
+    }
+
+    // Connection status id must be the same as the active network id
+    if (m_active_network_id != connection_status.id) {
+        LOG(ERROR) << "Network id mismatch: m_active_network_id(" << m_active_network_id << ") != "
+                   << "connection_status.id(" << connection_status.id << ")";
+        return false;
+    }
+
+    const std::string cmd = "REMOVE_NETWORK " + std::to_string(m_active_network_id);
+    if (!wpa_ctrl_send_msg(cmd)) {
+        LOG(ERROR) << "WPA command '" << cmd << "' failed";
+    }
+
+    // Clear state
+    m_active_ssid       = "";
+    m_active_bssid      = "";
+    m_active_pass       = "";
+    m_active_channel    = 0;
+    m_active_network_id = -1;
+
+    return true;
+}
+
+bool sta_wlan_hal_nl80211::roam(const sMacAddr &bssid, uint8_t channel)
+{
+    if (m_active_network_id == -1) {
+        LOG(ERROR) << "Incorrect active network " << m_active_network_id;
+        return false;
+    }
+
+    if (!is_connected()) {
+        LOG(ERROR) << get_iface_name() << " Not connected, can't roam";
+        return false;
+    }
+
+    auto bssid_str = tlvf::mac_to_string(bssid);
+
+    const std::string cmd = "ROAM " + bssid_str;
+    if (!wpa_ctrl_send_msg(cmd)) {
+        LOG(ERROR) << "WPA command '" << cmd << "' failed";
+        ;
+        return false;
+    }
+
+    auto freq = son::wireless_utils::channel_to_freq(int(channel));
+    if (!set_network(m_active_network_id, "freq_list", std::to_string(freq))) {
+        LOG(ERROR) << "Failed setting frequency on iface " << get_iface_name();
+        return false;
+    }
```

Comment:
I think this should be done *before* the ROAM, otherwise the channel parameter is kind of useless. Its purpose is to make sure that no full scan is needed to be able to switch APs, you can directly go to the correct channel. But then you have to update the channel list before calling ROAM, of course.

Reply:
Moved before ROAM